TFUtil.py

```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
+    labels_flat = tf.reshape(labels, [-1])
+
+    if sampled_values is None:
+      sampled_values = tf.random.log_uniform_candidate_sampler(
+          true_classes=labels,
+          num_true=num_true,
+          num_sampled=num_sampled,
+          unique=True,
+          range_max=num_classes,
+          seed=seed)
+
+    sampled, true_expected_count, sampled_expected_count = (
+        tf.stop_gradient(s) for s in sampled_values)
+    sampled = tf.cast(sampled, tf.int64)
+
+    all_ids = tf.concat([labels_flat, sampled], 0)
+
+    all_w = tf.nn.embedding_lookup(
+        weights, all_ids, partition_strategy=partition_strategy)
+    if all_w.dtype != inputs.dtype:
+      all_w = tf.cast(all_w, inputs.dtype)
+
+    true_w = tf.slice(all_w,
+                      [0, 0],
+                      tf.stack([tf.shape(labels_flat)[0], -1]))
+
+    sampled_w = tf.slice(
+        all_w, tf.stack([tf.shape(labels_flat)[0], 0]), [-1, -1])
+    sampled_logits = tf.matmul(inputs, sampled_w, transpose_b=True)
+
+    all_b = tf.nn.embedding_lookup(
+        biases, all_ids, partition_strategy=partition_strategy)
+    if all_b.dtype != inputs.dtype:
+      all_b = tf.cast(all_b, inputs.dtype)
+    true_b = tf.slice(all_b, [0], tf.shape(labels_flat))
+    sampled_b = tf.slice(all_b, tf.shape(labels_flat), [-1])
+
+    dim = tf.shape(true_w)[1:2]
+    new_true_w_shape = tf.concat([[-1, num_true], dim], 0)
+    row_wise_dots = tf.multiply(
+        tf.expand_dims(inputs, 1),
+        tf.reshape(true_w, new_true_w_shape))
+    dots_as_matrix = tf.reshape(row_wise_dots,
+                                tf.concat([[-1], dim], 0))
+    true_logits = tf.reshape(sum_rows(dots_as_matrix), [-1, num_true])
+    true_b = tf.reshape(true_b, [-1, num_true])
+    true_logits += true_b
+    sampled_logits += sampled_b
+
+    if remove_accidental_hits:
+      acc_hits = tf.random.compute_accidental_hits(
+          labels, sampled, num_true=num_true)
+      acc_indices, acc_ids, acc_weights = acc_hits
+
+      acc_indices_2d = tf.reshape(acc_indices, [-1, 1])
+      acc_ids_2d_int32 = tf.reshape(f.cast(acc_ids, tf.int32), [-1, 1])
+      sparse_indices = tf.concat([acc_indices_2d, acc_ids_2d_int32], 1,
+                                 "sparse_indices")
+      sampled_logits_shape = tf.concat(
+          [tf.shape(labels)[:1],
+           tf.expand_dims(num_sampled, 0)], 0)
+      if sampled_logits.dtype != acc_weights.dtype:
+        acc_weights = tf.cast(acc_weights, sampled_logits.dtype)
+      sampled_logits += tf.sparse_to_dense(
+          sparse_indices,
+          sampled_logits_shape,
+          acc_weights,
+          default_value=0.0,
+          validate_indices=False)
+
+    if subtract_log_q:
+      true_logits -= tf.log(true_expected_count)
+      sampled_logits -= tf.log(sampled_expected_count)
+
+    out_logits = tf.concat([true_logits, sampled_logits], 1)
+
+    out_labels = tf.concat([
+        tf.ones_like(true_logits) / num_true,
```

Comment:
Use `//` for integer division.

Reply:
I think this should be correct as is. If num_true > 1 and we do `floordiv` then the results should be `0.0` which is not what we want. This is just to ensure that the per-example labels sum to `1.0`

Reply:
Ah, sorry, I got confused. I thought that are labels (because you call it `out_labels`), i.e. integers. But why do you call it "labels" then? Better rename it, so that it is more clear what it is.

Reply:
I'm not sure if "targets" is better... Maybe "soft targets"? Well, maybe good enough like that.